File: conformance/run-1/components/matrix.ts
import { handle, OperationError } from '../../../packages/flow-sdk/src/index'

await handle(async (run) => {
  const input = run.input as { case?: unknown }
  switch (input.case) {
    case 'fanout-65': {
      const calls = Array.from({ length: 65 }, (_, index) =>
        run.call({
          operationId: `fanout:${index + 1}`,
          slot: 'sink',
          input: { index },
        }),
      )
      await Promise.allSettled(calls)

      return { outcome: 'done', output: { settled: calls.length } }
    }
    case 'operation-identity': {
      const call = () =>
        run.call({
          operationId: 'shared:1',
          slot: 'sink',
          input: { value: 'same' },
        })
      const [first, second] = await Promise.all([call(), call()])
      const replay = await call()
      let conflict: string | null = null
      try {
        await run.call({
          operationId: 'shared:1',
          slot: 'sink',
          input: { value: 'different' },
        })
      } catch (error) {
        if (!(error instanceof OperationError)) throw error
        conflict = error.code
      }
      return { outcome: 'done', output: { first, second, replay, conflict } }
    }
    case 'cancel-shared-waiter': {
      const controller = new AbortController()
      const call = (signal?: AbortSignal) =>
        run.call(
          {
            operationId: 'shared-cancel:1',
            slot: 'sink',
            input: { value: 'shared' },
          },
          signal === undefined ? undefined : { signal },
        )
      const cancelled = call(controller.signal)
      const survivor = call()
      await run.call({
        operationId: 'release-shared-cancel:1',
        slot: 'control',
        input: null,
      })
      controller.abort()
      let cancellation: string | null = null
      try {
        await cancelled
      } catch (error) {
        if (!(error instanceof OperationError)) throw error
        cancellation = error.code
      }
      return {
        outcome: 'done',
        output: { cancellation, survivor: await survivor },
      }
    }
    case 'uncertain-replay': {
      const call = async (operationId: string) => {
        try {
          return await run.call({
            operationId,
            slot: 'sink',
            input: { value: 'uncertain' },
          })
        } catch (error) {
          if (!(error instanceof OperationError)) throw error
          return error.code
        }
      }
      const first = await call('uncertain:1')
      const replay = await call('uncertain:1')
      const fresh = await call('uncertain:2')
      return { outcome: 'done', output: { first, replay, fresh } }
    }
    case 'request-lifetime': {
      let accepted = 0
      let rejected: string | null = null
      for (let index = 1; index <= 65_537; index += 1) {
        try {
          await run.call({
            operationId: `lifetime:${index}`,
            slot: 'sink',
            input: null,
          })
          accepted += 1
        } catch (error) {
          if (!(error instanceof OperationError)) throw error
          rejected = error.code
        }
      }
      return { outcome: 'done', output: { accepted, rejected } }
    }
    case 'one-flow': {
      const child = await run.call({
        operationId: 'child:1',
        slot: 'child',
        input: null,
      })
      return { outcome: 'done', output: child }
    }
    case 'two-effects': {
      const first = await run.call({
        operationId: 'first:1',
        slot: 'sink',
        input: { sequence: 1 },
      })
      const second = await run.call({
        operationId: 'second:1',
        slot: 'sink',
        input: { sequence: 2 },
      })
      return { outcome: 'done', output: { first, second } }
    }
    case 'cancel-one-call': {
      const controller = new AbortController()
      const child = run.call(
        {
          operationId: 'cancelled-child:1',
          slot: 'child',
          input: null,
        },
        { signal: controller.signal },
      )
      await run.call({
        operationId: 'release-cancel:1',
        slot: 'control',
        input: null,
      })
      controller.abort()
      try {
        await child
        throw new Error('cancelled child unexpectedly completed')
      } catch (error) {
        if (!(error instanceof OperationError) || error.code !== 'CANCELLED') {
          throw error
        }
      }
      return { outcome: 'done', output: 'cancelled-locally' }
    }
    case 'abandoned-call': {
      // Abandon the operation without also creating a language-level
      // unhandled rejection when the SDK closes its owner.
      void run
        .call({
          operationId: 'abandoned-child:1',
          slot: 'child',
          input: null,
        })
        .catch(() => undefined)
      await run.call({
        operationId: 'release-abandon:1',
        slot: 'control',
        input: null,
      })
      return { outcome: 'done', output: 'must-not-succeed' }
    }
    default:
      return { outcome: 'done', output: null }
  }
})

File: conformance/run-1/components/matrix.py
import asyncio

from flowmd_sdk import OperationError, handle


async def run(context):
    case = context.input.get("case")
    if case == "fanout-65":
        calls = [
            asyncio.create_task(
                context.call_effect(
                    operation_id=f"fanout:{index + 1}",
                    slot="sink",
                    method="write",
                    input={"index": index},
                )
            )
            for index in range(65)
        ]
        await asyncio.gather(*calls, return_exceptions=True)
        return {"outcome": "done", "output": {"settled": len(calls)}}

    if case == "operation-identity":
        async def call():
            return await context.call_effect(
                operation_id="shared:1",
                slot="sink",
                method="write",
                input={"value": "same"},
            )

        first_task = asyncio.create_task(call())
        second_task = asyncio.create_task(call())
        first, second = await asyncio.gather(first_task, second_task)
        replay = await call()
        conflict = None
        try:
            await context.call_effect(
                operation_id="shared:1",
                slot="sink",
                method="write",
                input={"value": "different"},
            )
        except OperationError as error:
            conflict = error.code
        return {
            "outcome": "done",
            "output": {
                "first": first,
                "second": second,
                "replay": replay,
                "conflict": conflict,
            },
        }

    if case == "cancel-shared-waiter":
        async def call_shared():
            return await context.call_effect(
                operation_id="shared-cancel:1",
                slot="sink",
                method="write",
                input={"value": "shared"},
            )

        cancelled = asyncio.create_task(call_shared())
        await asyncio.sleep(0)
        survivor = asyncio.create_task(call_shared())
        await asyncio.sleep(0)
        await context.call_effect(
            operation_id="release-shared-cancel:1",
            slot="control",
            method="release",
            input=None,
        )
        cancelled.cancel()
        cancellation = None
        try:
            await cancelled
        except asyncio.CancelledError:
            cancellation = "CANCELLED"
        return {
            "outcome": "done",
            "output": {
                "cancellation": cancellation,
                "survivor": await survivor,
            },
        }

    if case == "uncertain-replay":
        async def call_uncertain(operation_id):
            try:
                return await context.call_effect(
                    operation_id=operation_id,
                    slot="sink",
                    method="write",
                    input={"value": "uncertain"},
                )
            except OperationError as error:
                return error.code

        first = await call_uncertain("uncertain:1")
        replay = await call_uncertain("uncertain:1")
        fresh = await call_uncertain("uncertain:2")
        return {
            "outcome": "done",
            "output": {"first": first, "replay": replay, "fresh": fresh},
        }

    if case == "request-lifetime":
        accepted = 0
        rejected = None
        for index in range(1, 65_538):
            try:
                await context.call_effect(
                    operation_id=f"lifetime:{index}",
                    slot="sink",
                    method="write",
                    input=None,
                )
                accepted += 1
            except OperationError as error:
                rejected = error.code
        return {
            "outcome": "done",
            "output": {"accepted": accepted, "rejected": rejected},
        }

    if case == "one-flow":
        child = await context.call_flow(
            operation_id="child:1",
            slot="child",
            input=None,
        )
        return {"outcome": "done", "output": child}

    if case == "two-effects":
        first = await context.call_effect(
            operation_id="first:1",
            slot="sink",
            method="write",
            input={"sequence": 1},
        )
        second = await context.call_effect(
            operation_id="second:1",
            slot="sink",
            method="write",
            input={"sequence": 2},
        )
        return {"outcome": "done", "output": {"first": first, "second": second}}

    if case == "cancel-one-call":
        child = asyncio.create_task(
            context.call_flow(
                operation_id="cancelled-child:1",
                slot="child",
                input=None,
            )
        )
        await context.call_effect(
            operation_id="release-cancel:1",
            slot="control",
            method="release",
            input=None,
        )
        child.cancel()
        try:
            await child
        except asyncio.CancelledError:
            pass
        else:
            raise AssertionError("cancelled child unexpectedly completed")
        return {"outcome": "done", "output": "cancelled-locally"}

    if case == "abandoned-call":
        asyncio.create_task(
            context.call_flow(
                operation_id="abandoned-child:1",
                slot="child",
                input=None,
            )
        )
        await context.call_effect(
            operation_id="release-abandon:1",
            slot="control",
            method="release",
            input=None,
        )
        return {"outcome": "done", "output": "must-not-succeed"}

    return {"outcome": "done", "output": None}


handle(run)
